cuda/gemv_benchmark.cu

```diff
@@ -0,0 +1,282 @@
+            },
+            ar_compute_error},
+        benchmark_info_t{
+            "GEMV Acc<fp64, fp32>",
+            [&](matrix_info m_info, matrix_info x_info, matrix_info res_info) {
+                acc_gemv<ar_type>(m_info, st_alpha, st_data.gpu_mtx_const(),
+                                  x_info, st_data.gpu_x_const(), res_info,
+                                  st_beta, st_data.gpu_res());
+            },
+            st_compute_error},
+        benchmark_info_t{
+            "GEMV Acc<fp32, fp32>",
+            [&](matrix_info m_info, matrix_info x_info, matrix_info res_info) {
+                acc_gemv<st_type>(m_info, st_alpha, st_data.gpu_mtx_const(),
+                                  x_info, st_data.gpu_x_const(), res_info,
+                                  st_beta, st_data.gpu_res());
+            },
+            st_compute_error},
+        benchmark_info_t{
+            "CUBLAS GEMV fp64",
+            [&](matrix_info m_info, matrix_info x_info, matrix_info res_info) {
+                cublas_gemv(cublasHandle.get(), m_info, ar_alpha,
+                            ar_data.gpu_mtx_const(), x_info,
+                            ar_data.gpu_x_const(), res_info, ar_beta,
+                            ar_data.gpu_res());
+            },
+            ar_compute_error},
+        benchmark_info_t{
+            "CUBLAS GEMV fp32",
+            [&](matrix_info m_info, matrix_info x_info, matrix_info res_info) {
+                cublas_gemv(cublasHandle.get(), m_info, st_alpha,
+                            st_data.gpu_mtx_const(), x_info,
+                            st_data.gpu_x_const(), res_info, st_beta,
+                            st_data.gpu_res());
+            },
+            st_compute_error},
+    };
+
+    std::cout << "Num rows";
+    for (const auto &info : benchmark_info) {
+        if (!measure_error) {
+            std::cout << DELIM << std::get<0>(info);
+        } else {
+            std::cout << DELIM << "Error " << std::get<0>(info);
+        }
+    }
+    std::cout << '\n';
+
+    std::cout.precision(16);
+    // showpos: show + sign for positive numbers
+    std::cout << std::scientific << std::showpos;
+    constexpr auto start = 100;
+    constexpr auto row_incr = start;
+    for (std::size_t num_rows = start; num_rows <= max_rows;
+         num_rows += row_incr) {
+        std::array<value_type, benchmark_num> local_res{};
+        const matrix_info m_info{{num_rows, num_rows}, max_rows};
+        const matrix_info x_info{{num_rows, 1}};
+        const matrix_info res_info{{num_rows, 1}};
+
+        if (measure_error) {
+            std::get<1>(benchmark_info[benchmark_reference])(m_info, x_info,
+                                                             res_info);
+            cpu_res_ref = ar_data.gpu_res_memory();
+            if (normalize_error) {
+                res_ref_norm = reduce<value_type>(
+                    res_info, cpu_res_ref.data(), [](ar_type a, ar_type b) {
```

Comment:
does the data() return cpu data? I thought it is on gpu

Reply:
`cpu_res_ref` is stored on the CPU. The line above: `cpu_res_ref = ar_data.gpu_res_memory();` copies the data from the GPU to the CPU (`operator=` os overloaded). I make it more clear by calling `copy_from`.

Reply:
I see, I do not notice the overload